contracts/state-validator/src/verifications/submit_block/global_state_v2.rs

```diff
@@ -0,0 +1,133 @@
+use alloc::collections::BTreeSet;
+use gw_utils::{
+    cells::rollup::MAX_ROLLUP_WITNESS_SIZE,
+    ckb_std::{
+        ckb_constants::Source,
+        debug,
+        syscalls::{load_witness, SysError},
+    },
+    error::Error,
+    gw_types::{
+        packed::{
+            GlobalState, LastFinalizedWithdrawal, RawL2BlockReader, ScriptVecReader,
+            WithdrawalRequestReader, WitnessArgsReader,
+        },
+        prelude::{Builder, Entity, Pack, Reader, Unpack},
+    },
+};
+
+use crate::verifications::finalize_withdrawal::last_finalized_withdrawal::LAST_FINALIZED_WITHDRAWAL_INDEX_ALL_WITHDRAWALS;
+
+pub struct GlobalStateV2Verifications {
+    pub check_no_output_withdrawal_cells: bool,
+    pub check_withdrawal_owner_lock_in_last_witness_type_out: bool,
+    pub check_prev_last_finalized_withdrawal_field_is_default: bool,
+}
+
+// NOTE: The only place that modify `last_finalized_withdrawal` in submit block is upgrade to
+// v2 process, so we don't need to check prev last finalized withdrawal field is default when
+// post global state is v2.
+impl GlobalStateV2Verifications {
+    pub fn from_post_global_state(post_global_state: &GlobalState) -> Result<Self, Error> {
+        let verifications = match post_global_state.version_u8() {
+            0 | 1 => GlobalStateV2Verifications {
+                check_no_output_withdrawal_cells: false,
+                check_withdrawal_owner_lock_in_last_witness_type_out: false,
+                check_prev_last_finalized_withdrawal_field_is_default: true,
+            },
+            2 => GlobalStateV2Verifications {
+                check_no_output_withdrawal_cells: true,
+                check_withdrawal_owner_lock_in_last_witness_type_out: true,
+                check_prev_last_finalized_withdrawal_field_is_default: false,
+            },
+            ver => {
+                debug!("invalid global state version {}", ver);
+                return Err(Error::InvalidGlobalStateVersion);
+            }
+        };
+
+        Ok(verifications)
+    }
+
+    pub fn check_withdrawal_owner_lock<'a>(
+        withdrawals: &[WithdrawalRequestReader<'a>],
+    ) -> Result<(), Error> {
```

Comment:
This structure will/may be used by some other code, e.g. selection of finality rules. And because the structure is not strongly relevant to the verification, I hope that the verification code could be separated from this common structure.

Reply:
And I would like to refactor this structure as follows:

```rust
pub struct ForkSwitch {
    post_version: u8
}

impl ForkSwitch {
    fn new(post_version: u8) -> Self {
        assert!(post_version <= 2);
        Self { post_version }
    }

    fn check_no_output_withdrawal_cells(&self) -> bool {
        self.post_version >= 2
    }

    fn check_withdrawal_owner_lock_in_last_witness_type_out(&self) -> bool {
        self.post_version >= 2
    }

    fn check_prev_last_finalized_withdrawal_field_is_default(&self) -> bool {
        self.post_version <= 1
    }
}
```

Reply:
And what about making `ForkSwitch` be one field of `BlockContext`? version is also one context of block.

Reply:
Ok, make sense. Done https://github.com/nervosnetwork/godwoken-scripts/pull/126/commits/88adb634f2f64442b0c77771e8683bf1fe12901a